### ISSUE TITLE
Connection timeout: pool is probably full.

### ISSUE THREAD

**Issue by shuo-tian** [knex/knex#6034]
# Environment

Knex version: 3.1.0
Database + version:  `PostgreSQL 13.13 (Ubuntu 13.13-1.pgdg20.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0, 64-bit`
OS: Windows 10
Node: 16.20.2
pg: 8.11.3

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do
	I'm running the back end of a web app. It should start without an error

2. Error message
![image](https://github.com/knex/knex/assets/120124789/51664de5-a8ba-4cf3-8313-c9fc283fe019)
```
"Error while establishing postgresql connection:  Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
 
   KnexTimeoutError : Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
```

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to MySQL or PostgreSQL, then single file example which initializes
   needed data and demonstrates the problem.

I don't have one for now. I can create a sandbox later if needed.

# configs:
```js
module.exports = {
  client: "postgresql",
  connection: {
    connectionString: process.env.DATABASE_URL,
    ssl: { rejectUnauthorized: false },
  },
  acquireConnectionTimeout: 1000000,
  "pool": {
    "min": 2,
    "max": 6,
    "createTimeoutMillis": 3000,
    "acquireTimeoutMillis": 30000,
    "idleTimeoutMillis": 30000,
    "reapIntervalMillis": 1000,
    "createRetryIntervalMillis": 100,
    "propagateCreateError": true
  },
  migrations: {
    tableName: "migrations"
  },  
};
```

I've seen the same error with Prisma, which is [documented here](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections/connection-pool). I wonder if something similar can be done here.

**Comment by jarzt-pham** [comment]
I have the same problem

**Comment by rluvaton** [comment]
Can you please provide a minimal reproduction repository

**Comment by shuo-tian** [comment]
> Can you please provide a minimal reproduction repository

No, as I can't reproduce it consistently. Sometimes re-running `npm run dev` after seeing this error fixes it.

**Comment by rluvaton** [comment]
I don't know how I can help without a minimal reproduction repository as it is probably a local setup, when it happens try to connect debug the db connections

**Comment by elindgren** [comment]
I'm experiencing this problem as well I have not been able to replicate the issue locally, but it occurs in both our staging and production environments, typically days/weeks after a new deploy. I suspect there is an issue which causes the pool to slowly fill up, which eventually leads to these issues.

**Comment by shujer** [comment]
This problem appears after running for a period of time, and we almost encounter it weekly. it's hard to reproduce locally, maybe some usage or configuration will cause the pool full and connection timeout. we already set pool max size to 200 but not working.

**Comment by mercmobily** [comment]
This is something we really should look at once we have tamed the issue list. Bugs will have priority anyway... this one is gonna be "fun"!

**Comment by mercmobily** [comment]
This connection pool exhaustion issue is being addressed through our new lazy transaction implementation!

We're implementing lazy transactions (#6258) that defer connection acquisition until the first query is actually executed. This will prevent the "pool is probably full" errors that occur when transactions are created but never used - a common pattern that unnecessarily consumes connections.

This is part of a comprehensive effort (#6257) to modernise Knex's transaction and connection management based on the original vision from PR #1659.

With lazy transactions:
- Creating a transaction won't immediately take a connection from the pool
- Unused transactions won't cause pool exhaustion
- Connection timeouts will be dramatically reduced

Thank you for reporting this issue. Closing in favour of the implementation that will solve the root cause.
